redwood-lazylayout-widget/src/commonMain/kotlin/app/cash/redwood/lazylayout/widget/WindowedLazyList.kt

```diff
@@ -30,10 +30,16 @@ public abstract class WindowedLazyList<W : Any>(
   }
 
   protected fun updateViewport(firstVisibleItemIndex: Int, lastVisibleItemIndex: Int) {
-    if (firstVisibleItemIndex != this.firstVisibleItemIndex || lastVisibleItemIndex != this.lastVisibleItemIndex) {
-      this.firstVisibleItemIndex = firstVisibleItemIndex
-      this.lastVisibleItemIndex = lastVisibleItemIndex
-      onViewportChanged?.invoke(firstVisibleItemIndex, lastVisibleItemIndex)
+    // Paginate the results so `onViewportChanged` isn't blasted with a bunch of updates. This is
+    // particularly important when using `LazyList` in Treehouse (compared to plain Redwood), as
+    // the serialization cost between the host-guest bridge can be expensive.
+    val viewportItemCount = lastVisibleItemIndex - firstVisibleItemIndex
+    val firstPagedItemIndex = (((firstVisibleItemIndex / viewportItemCount) * viewportItemCount) - viewportItemCount).coerceAtLeast(0)
+    val lastPagedItemIndex = ((lastVisibleItemIndex / viewportItemCount) * viewportItemCount) + viewportItemCount
```

Comment:
Now that we have logic in here, can we write tests for this class?

I'm specifically worried that this algorithm is still a bit off in terms of what it requests. Today, if we're showing 19..29 the count would be 10, the first paged index would be 0 (19/10=1, 1*10=10, 10-10=0) and the last paged index would be 30 (29/10=2, 2*10=20, 20+10=30). This means moving down the list will cause placeholders to be visible since there's only a single item offscreen in that direction, whereas there's 19 offscreen items in the other direction. A quick fix would be to `+1` after the division.

Reply:
Good point. I just added a basic test case for this class to ensure that it's working. As you described, the logic was slightly off, so it's been since updated.